### PR TITLE
DUOS-1215[risk=no]Add missing final vote info to Review_Results page

### PR DESCRIPTION
**Pull request by raejohanek**
SCOPE:
Currently, the UI calls getAccessElectionReviewByReferenceId to render the votes on the access review and review result page, with the boolean flag true meaning isfinalaccess. Due to this, the call in consent was filtering for only votes with FINAL type. These are all chair votes. This means that we are not receiving the votes for the members. 
This PR removes that boolean flag.
By switching to use this DAO call that has no boolean flag, we receive the DAC votes and the FINAL votes. This update is needed for this PR so that we can display both member and chair votes distinctly on the review page. Currently, the pages say "How members voted" but only show chair votes, so this will allow the page to be more accurate. 

DO NOT MERGE WITHOUT DUOS-1215 on UI

ADDRESSES:
https://broadworkbench.atlassian.net/browse/DUOS-1215

Have you read [CONTRIBUTING.md](../CONTRIBUTING.md) lately? If not, do that first.

I, the developer opening this PR, do solemnly pinky swear that:

- [ ] I've followed [the instructions](https://github.com/DataBiosphere/consent/blob/develop/CONTRIBUTING.md#api-changes) if I've made any changes to the API, _especially_ if they're breaking changes
- [ ] I've updated the RC_XXX release ticket with any manual steps required to release this change
- [ ] I've updated the [FISMA documentation](https://github.com/DataBiosphere/consent/blob/develop/CONTRIBUTING.md#fisma-documentation-changes) if I've made any security-related changes, including auth, encryption, or auditing
- [ ] I've updated Swagger to reflect any API changes

In all cases:

- [ ] Get two thumbsworth of review and PO signoff if necessary
- [ ] Verify all tests go green
- [ ] Squash and merge; you can delete your branch after this **unless it's for a hotfix**. In that case, don't delete it!
- [ ] Test this change deployed correctly and works on dev environment after deployment

### CODE REVIEW

**Comment by rushtong** [comment]
The underlying premise of this PR is not accurate. The UI does query for all votes on an election. I think the right approach here is to update the UI to make use of the votes it is already receiving. See screenshot for an example where you can see that we're getting votes of all types for the election.

Screen Shot 2021-04-29 at 6 35 03 AM

**Comment by raejohanek** [comment]
> The underlying premise of this PR is not accurate. The UI does query for all votes on an election. I think the right approach here is to update the UI to make use of the votes it is already receiving. See screenshot for an example where you can see that we're getting votes of all types for the election.
> 
> Screen Shot 2021-04-29 at 6 35 03 AM

To see the votes I am talking about go to to review results page and look at the accessElectionReview that is saved in state, this contains the votes that are being displayed in the vote summaries and this includes only the chairperson type FINAL votes, meaning all member votes are being ignored, and never get displayed. This PR attempts to correct that functionality by receiving DAC (member) and FINAL (chair) votes. If we want to get away from using accessElectionReview altogether, like you were saying in stand up, then that is a larger issue outside the scope of this PR. But on that topic, the only information the electionreview objects are distinctly used for are the reviewvotes which contain user names. So, if we want to eliminate this object altogether, then that is the one thing we will need to implement elsewhere.

File: src/main/java/org/broadinstitute/consent/http/resources/ElectionReviewResource.java [hunk 1]
```diff
@@ -66,7 +66,7 @@ public ElectionReview getElectionReviewByElectionId(@PathParam("electionId") Int
     @Path("access/{electionId}")
     @Produces("application/json")
     @RolesAllowed({ADMIN, MEMBER, CHAIRPERSON, ALUMNI})
-    public ElectionReview getAccessElectionReviewByReferenceId(@PathParam("electionId") Integer electionId, @QueryParam("isFinalAccess") Boolean isFinalAccess) {
+    public ElectionReview getAccessElectionReviewByReferenceId(@PathParam("electionId") Integer electionId) {
```
Comment (rushtong) [inline src/main/java/org/broadinstitute/consent/http/resources/ElectionReviewResource.java]: This changes the api signature which would require a change to the api-docs as well. This method is called from the UI and provides the `isFinalAccess` as well so this PR will require a companion on the UI side.

File: src/main/java/org/broadinstitute/consent/http/resources/ElectionReviewResource.java [hunk 2]
```diff
@@ -75,7 +75,7 @@ public ElectionReview getAccessElectionReviewByReferenceId(@PathParam("electionI
             dataSetId.addAll(dar.getData().getDatasetIds());
         }
         Consent consent = consentService.getConsentFromDatasetID(dataSetId.get(0));
-        ElectionReview accessElectionReview = service.describeElectionReviewByElectionId(electionId, isFinalAccess);
+        ElectionReview accessElectionReview = service.describeElectionReviewByElectionId(electionId, null);
```
Comment (rushtong) [inline src/main/java/org/broadinstitute/consent/http/resources/ElectionReviewResource.java]: I think using null for an argument as a generic anti-pattern. We should strive to avoid it if at all possible. That said, it isn't always possible and sometimes we need to handle legacy code this way.

File: src/main/java/org/broadinstitute/consent/http/service/ReviewResultsService.java [hunk 3]
```diff
@@ -48,7 +48,10 @@ public ElectionReview describeElectionReviewByElectionId(Integer electionId, Boo
             review = new ElectionReview();
             review.setElection(election);
             Consent consent = consentDAO.findConsentById(review.getElection().getReferenceId());
-            List<ElectionReviewVote> rVotes = (isFinalAccess == null || isFinalAccess == false) ? voteDAO.findElectionReviewVotesByElectionId(electionId, VoteType.DAC.getValue()) :  voteDAO.findElectionReviewVotesByElectionId(electionId, VoteType.FINAL.getValue());
+            List<ElectionReviewVote> rVotes = isFinalAccess == null ? 
+              voteDAO.findElectionReviewVotesByElectionId(electionId)
+              : isFinalAccess == false ? voteDAO.findElectionReviewVotesByElectionId(electionId, VoteType.DAC.getValue())
+                : voteDAO.findElectionReviewVotesByElectionId(electionId, VoteType.FINAL.getValue());
```
Comment (rushtong) [inline src/main/java/org/broadinstitute/consent/http/service/ReviewResultsService.java]: The nested ternaries here add a significant degree of complexity that's hard to follow. I'd like to see this broken down so the logic can be read more quickly and cleanly. 

What I think makes this more complex is that we now have 3 operational states of `isFinalAccess`, null, true, or false, and each has different semantics. Using null this way is definitely an anti-pattern. A better approach would be to create an overloaded method with the arguments you want, and then calling that one from the resource class, it might look like this ... 

```
    public ElectionReview describeElectionReviewByElectionId(Integer electionId) {
        ElectionReview review = null;
        Election election = electionDAO.findElectionWithFinalVoteById(electionId);
        if(election != null){
            review = new ElectionReview();
            review.setElection(election);
            Consent consent = consentDAO.findConsentById(review.getElection().getReferenceId());
            List<ElectionReviewVote> rVotes = voteDAO.findElectionReviewVotesByElectionId(electionId);
            review.setReviewVote(rVotes);
            review.setConsent(consent);
        }
        return review;
    }
```

**Review by rushtong — CHANGES_REQUESTED** [review]
See comments inline. Let's make sure this is well tested against the UI.

File: src/main/java/org/broadinstitute/consent/http/service/ReviewResultsService.java (hunk 3 above)
Comment (raejohanek) [reply src/main/java/org/broadinstitute/consent/http/service/ReviewResultsService.java]: yeah good idea I agree and if we want to make this same change for the rp election, so that we are receiving both member and chair votes, then we could get rid of the isfinalaccess boolean flag altogether and only have this version

Comment (rushtong) [reply src/main/java/org/broadinstitute/consent/http/service/ReviewResultsService.java]: That also makes a lot of sense. I'm a little worried that would introduce unexpected regressions, but it's probably worth the effort to do that now rather than later.

**Review by rushtong — APPROVED** [review]
Pending a successful test run, 👍🏽. Try re-running the failing job, I believe it's a flaky test that might have been made more flaky with a recent library update.

**Comment by raejohanek** [comment]
> Pending a successful test run, 👍🏽. Try re-running the failing job, I believe it's a flaky test that might have been made more flaky with a recent library update.

After testing this through the UI, I made a change to the voteDAO, I removed the filter on vote type != 'chairperson'. On the access election this filter does not cause any problems because every chairperson has a chairperson vote, and a final vote. But on the rp election, the chairs don't have a final vote. So this filter is causing the "How committee chairs voted:" section for question 2 not to populate. If we don't want to remove this filter for some reason (I don't know why it is there in the first place) we could simply replace it with type != 'final'

**Comment by rushtong** [comment]
@raejohanek I'm not following this conversation well, I think we need to talk through what's going on.

> After testing this through the UI, I made a change to the voteDAO, I removed the filter on vote type != 'chairperson'. On the access election this filter does not cause any problems because every chairperson has a chairperson vote, and a final vote. But on the rp election, the chairs don't have a final vote. So this filter is causing the "How committee chairs voted:" section for question 2 not to populate. If we don't want to remove this filter for some reason (I don't know why it is there in the first place) we could simply replace it with type != 'final'

File: src/main/java/org/broadinstitute/consent/http/db/VoteDAO.java [hunk 4]
```diff
@@ -28,6 +28,11 @@
     @UseRowMapper(ElectionReviewVoteMapper.class)
     List<ElectionReviewVote> findElectionReviewVotesByElectionId(@Bind("electionId") Integer electionId);
 
+    @SqlQuery("select v.*, u.email, u.displayName from vote v inner join election on election.electionId = v.electionId inner join dacuser u on u.dacUserId = v.dacUserId "
+            + "where election.electionId = :electionId")
+    @UseRowMapper(ElectionReviewVoteMapper.class)
+    List<ElectionReviewVote> findAllElectionReviewVotesByElectionId(@Bind("electionId") Integer electionId);
+
```
Comment (raejohanek) [inline src/main/java/org/broadinstitute/consent/http/db/VoteDAO.java:37]: made new DAO call instead of modifying

File: src/main/java/org/broadinstitute/consent/http/resources/ElectionReviewResource.java (hunk 1 above)
Comment (rushtong) [reply src/main/java/org/broadinstitute/consent/http/resources/ElectionReviewResource.java]: Here and below, to make this more backwards compatible, leave the query param in place so that the method signature doesn't change, but instead just ignore the query param. That way, UI code can change separately from this and either call this endpoint with or without the query param. In a follow-up PR, we can remove the unused query param. That maintains backward compatibility and allows for the back end and the front end to be released independently.

**Review by rushtong — CHANGES_REQUESTED** [review]
One requested change to keep the original method signatures in place. After that, 👍🏽

File: src/main/resources/assets/api-docs.yaml [hunk 5]
```diff
@@ -2819,7 +2819,7 @@ paths:
                 $ref: '#/components/schemas/ElectionReview'
   /api/electionReview/rp/{electionId}:
     get:
-      summary: describeElectionById
+      summary: getRPElectionReviewByReferenceId
```
Comment (rushtong) [inline src/main/resources/assets/api-docs.yaml]: This should be a plain text shortened version of the description, not the method name. The method-name pattern was used by Belatrix, but we should update these to be more descriptive so the swagger UI is easier to read. Don't hold up the PR on this, so feel free to merge if you do update these text summaries.

File: src/main/resources/assets/api-docs.yaml [hunk 6]
```diff
@@ -2773,7 +2773,7 @@ paths:
                 $ref: '#/components/schemas/ElectionReview'
   /api/electionReview/access/{electionId}:
     get:
-      summary: describeElectionById
+      summary: getAccessElectionReviewByReferenceId
```
Comment (rushtong) [inline src/main/resources/assets/api-docs.yaml]: Same description comment here as below.

**Review by rushtong — APPROVED** [review]
Two minor suggestions inline, but then 👍🏽
